File: 백준/Silver/2468. 안전 영역/안전 영역.js
// 연결 찾기
const rl = require("readline").createInterface({
  input: process.stdin,
  output: process.stdout,
});

let input = [];

const getSafeCount = (n, map, threshold) => {
  let count = 0;
  const m = map.map((row) => row.slice());

  const dirs = [
    [0, 1],
    [0, -1],
    [1, 0],
    [-1, 0],
  ];

  const search = ([y, x]) => {
    if (y < 0 || x < 0 || y >= n || x >= n || m[y][x] <= threshold) return;

    m[y][x] = threshold;
    for (const [dy, dx] of dirs) search([dy + y, dx + x]);
  };

  for (let i = 0; i < m.length; i++)
    for (let j = 0; j < m.length; j++) {
      if (m[i][j] > threshold) {
        search([i, j]);
        count++;
      }
    }

  return count;
};

rl.on("line", (line) => input.push(line)).on("close", function () {
  const n = +input[0];
  let max = 0;
  let min = 100;
  const map = input.slice(1).map((l) =>
    l.split(" ").map((v) => {
      max = Math.max(max, +v);
      min = Math.min(min, +v);
      return +v;
    })
  );

  let res = 0;
  for (let i = min - 1; i <= max; i++) {
    res = Math.max(res, getSafeCount(n, map, i));
  }

  console.log(res);
});

File: 백준/Silver/2468. 안전 영역/안전 영역.py
import sys

sys.setrecursionlimit(10**6)

n = int(sys.stdin.readline())
maps = []

max_height = 1
min_height = 100

for _ in range(n):
    line = []
    for v in list(map(int, sys.stdin.readline().split(" "))):
        max_height = max(v, max_height)
        min_height = min(v, min_height)
        line.append(v)
    maps.append(line)

# 다 안잠겼을때 => 1 (최소는 1)
max_area = 1

adjs = [(0, 1), (0, -1), (1, 0), (-1, 0)]

for cur_h in range(min_height, max_height):
    # 안전공간 개수 찾기
    safe_area = 0
    safe_map = [[h > cur_h for h in row] for row in maps]

    def adj_safe_remove(y, x):
        for dy, dx in adjs:
            ny = dy + y
            nx = dx + x
            if ny < 0 or nx < 0 or ny >= n or nx >= n or safe_map[ny][nx] == False:
                continue
            safe_map[ny][nx] = False
            adj_safe_remove(ny, nx)

    for y in range(n):
        for x in range(n):
            if safe_map[y][x]:
                safe_area += 1
                safe_map[y][x] = False
                adj_safe_remove(y, x)

    max_area = max(safe_area, max_area)


print(max_area)
